synapse_antispam/mjolnir/ban_list.py

```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from .list_rule import ListRule, ALL_RULE_TYPES, USER_RULE_TYPES, SERVER_RULE_TYPES, ROOM_RULE_TYPES
+from twisted.internet import defer
+
+logger = logging.getLogger("synapse.contrib." + __name__)
+
+class BanList(object):
+    def __init__(self, api, room_id):
+        self.api = api
+        self.room_id = room_id
+        self.server_rules = []
+        self.user_rules = []
+        self.room_rules = []
+        self.build()
```

Comment:
Note that this won't run synchronously, so the fields won't necessarily be up to date when things get called. I think we would probably rather want to have async getters here that fetched the data on demand (relying on caches to make it cheap), which has the added benefit of meaning we don't need to restart synapse to pick up a change in ban lists

Reply:
we also rebuild lists over time as we see changes, so a restart shouldn't be required. This also isn't meant as a perfect solution, just something to catch 99.9% of the noise - a few seconds on startup waiting for a query isn't the end of the world imo.

We can make it more lazy, though it starts to affect message send times. The behaviour here is sacrificing a few events getting through in favour of being faster.